### DISCUSSION TITLE
Possible BUG - Priority in Tasks not Considered when Task finishes.

### DISCUSSION THREAD

**Discussion by christianbrugger** [apache/airflow#26821] [Q&A]
I am having a hard to reproduce issue / bug, so I am asking here for some input on how to proceed from here. Am I getting something wrong about airflow? Is this not supported? Have I configured something wrong? Or is this an bug?

### The issue
I noticed situations on our installation where low-priority tasks are sometimes scheduled while high-priority tasks are ready to be scheduled. Our DAGs are all linear and we use WeightRule.UPSTREAM, meaning the priority is increasing further down the tasks. That means once a DAG starts it should finish completely. All DAGs are running in a Pool with a slot limit of 5. We usually schedule our DAGs manually and they are long running for 1 - 5 hours. So it is really obvious to us when we see, besides the 5 DAGs running, one or two DAGs with only a little progress.

This is bad for us, as people are waiting for the result of the complete DAG and when other DAGs are scheduled in the meantime, it makes them wait longer unnecessarily.

### Our system

We run Airflow 2.4.1 on Python 3.8. We run everything on one small server with 2 CPU cores. We use a local MySQL database, have 1 scheduler, 1 local executor and 1 webserver worker running. We have about 100 DAGs. A few run every 15 minutes, doing database work. While the other we schedule manually. The actual processing is done on other servers through docker. 

### How I tried to reproduce it so far 
I defined two simple DAGs basically just waiting. They are both part of a pool with 1 slot. 

DAG A Tasks | Priority | Runtime
-- | -- | --
A1 | 1 | 6 seconds
A2 | 2 | 1 minute
A3 | 3 | 1 minute
A4 | 4 | 1 minute

DAG B Tasks | Priority | Runtime
-- | -- | --
B1 | 1 | 6 seconds
B2 | 2 | 1 minute
B3 | 3 | 1 minute
B4 | 4 | 1 minute

Now I manually schedule first a run for DAG A, and then a few seconds later for DAG B. Over and over again.

More than 90% of the time DAG B is on status scheduled until DAG A is fully processed, and then starts running. The schedule being 
```A1, A2, A3, A4, B1, B2, B3, B4```

However in about 5-10% of the cases, DAG B starts running before A is finished. With the schedule being for example: 
```A1, B1, A2, A3, A4, B2, B3, B4```
```A1, A2, B1, A3, A4, B2, B3, B4```
```A1, A2, B1, B2, A3, A4, B3, B4```

### What I expect to happen

I know there are different run states tasks can be, as shown in the [Task Lifecycle Documentation](https://airflow.apache.org/docs/apache-airflow/1.10.10/concepts.html#task-lifecycle) After A1 finishes, A2 is in "No status" and B1 is in "Scheduled". Still I would argue the following to be a guarantee:

- A1 task is marked successful before a pool slot is freed.
- The next time the scheduler runs it first checks all Tasks with "No status" that now can be scheduled.
- At this point A2 should be allowed to be scheduled and marked as such.
- Then the scheduler should select A2 as the next running task.

This is however not what is happening.

### What I tried so far

I looked through the configuration to look for any timeouts that are set after a task is finished. I couldn't find any. I looked in the config file how long the DAG script runs. It runs 2.5 seconds to generate about 90 DAGs.

I also disabled the "mini-scheduler" by setting `schedule_after_task_execution = False`. One might expect the problem getting worse, but it actually got much better. Still the issue occurs.

I removed some of the Sensors (poke mode) with normal PythonOperators, as it often happened after those, and I thought maybe there are treated differently. Not much changed.

I am now a bit out of luck. The next thing would be to increase the logging of the scheduler, add maybe some custom message. However as this is a rare occurring bug, every hour it happens once. I am trying to see if someone of you has any idea, of what to do next?

### Workaround

The only workaround I see so far, is to tell our Users to only start 5 DAGs at the same time, and then run more after they finished, when they want the results quickly.

**Comment by potiuk** [comment]
Thanks for so detailed report and sorry for delay in responding ( I was at ApacheCon).

I think it needs more looking at @ashb and @ephraimbuddy have been lookig into scheduling and maybe they will have a clue.

By a brief look, It might be a problem (`bug` or `feature` depending how you look at it). 

 I think the priority including upstream might only be taken into account only at the stage where tasks are moved from "schediuled" to "queued" state. The `total_priority_weight` is calculated when the taks is already scheduled  - and the way scheduler works in Airflow  2 is that it will schedule tasks in batches - and I believe there `total_priority_weight` is not used to select which tasks to schedule. 

So it might be scheduler will schedule some `lower total priority tasks` first and they will be started first before the `higher total priority` tasks get scheduled. That's my educated guess for now at least.

I am now going for holidays, so I might not be able to answer quickly, but possibly the more experts in this area can help.
